functions/shiritori/index.js

```diff
@@ -14,116 +14,104 @@
-    const begin = wordHira.slice(0, k.length)
-    if (begin === k) {
+    if ((chain === undefined) || (chain.length === 0)) {
       return true
     }
-  }
-  return false
-}
 
-// 名詞からしりとりのひらがなを選ぶ。
-exports.kanas = (word) => new Set((() => {
-  if (kuroshiroLoaded) {
-    word = kuroshiro.toHiragana(word)
-  }
+    // 使った名詞をチェックする。
+    const chainHira = chain.map(wanakana.toHiragana)
+    if (chainHira.indexOf(wordHira, 0) !== -1) {
+      throw new Bad('already used')
+    }
 
-  const wordKana = wanakana.toKatakana(word)
-  const wordHira = wanakana.toHiragana(wordKana)
-  const kk = wordKana[word.length - 1]
-  const hk = wordHira[wordHira.length - 1]
+    // しりとりの最初の文字をチェックする。
+    return exports.kanas(chain[0]).then(validKanas => {
+      for (const k of validKanas) {
+        const begin = wordHira.slice(0, k.length)
+        if (begin === k) {
+          return true
+        }
+      }
+      throw new Bad('文字 does not match')
+    })
+  })
+})
 
-  if (hk === 'ん') {
-    return []
-  }
+// 名詞からしりとりのひらがなを選ぶ。
+exports.kanas = word => loaded
+  .then(() => kuroshiro.convert(word, { to: 'hiragana' }))
+  .then((word) => {
+    const wordKana = wanakana.toKatakana(word)
+    const wordHira = wanakana.toHiragana(wordKana)
+    const kk = wordKana[wordKana.length - 1]
+    const hk = wordHira[wordHira.length - 1]
+
+    if (hk === 'ん') {
+      return []
+    }
 
-  if ((kk === 'ー') && (wordKana.length > 1)) {
-    return [
-      wordHira[wordHira.length - 1],
-      wordHira[wordHira.length - 2]
-    ]
-  }
+    if ((kk === 'ー') && (wordKana.length > 1)) {
+      return [
+        wordHira[wordHira.length - 1],
+        wordHira[wordHira.length - 2]
+      ]
+    }
 
-  if (smallKanas.includes(hk)) {
+    if (smallKanas.includes(hk)) {
+      return [
+        String.fromCharCode(hk.charCodeAt(0) + 1),
+        wordHira.slice(-2)
+      ]
+    }
     return [
-      String.fromCharCode(hk.charCodeAt(0) + 1),
-      wordHira.slice(-2)
+      wordHira[wordHira.length - 1]
     ]
-  }
-
-  return [
-    wordHira[wordHira.length - 1]
-  ]
-})())
+  })
+  .then(result => new Set(result))
 
 // しりとりのゲームループ。
-exports.interact = (dict, word, chain) => new Promise((resolve, reject) => {
-  const next = exports.kanas(word)
-  if ((next.size === 0) || !exports.check(word, chain)) {
-    return reject({loose: true})
-  }
-  const key = next.values().next().value
-  dict(key).then((words) => {
-    const unused = []
-    if (words) {
-      for (const k of Object.keys(words)) {
-        if (!chain.includes(words[k])) {
-          unused.push(k)
-        }
+exports.interact = (dict, word, chain) => exports.check(word, chain)
+    .then(() => exports.kanas(word))
+    .then(validKanas => dict(validKanas.values().next().value))
+    .then(words => {
+      const unused = words ? Object.keys(words).filter(k => !chain.includes(words[k])) : []
+      if (unused.length === 0) {
+        throw new Win('no more unused word remaining in dictionary')
```

Comment:
It feels weird to throw an error on a success condition. Maybe you could call it something more like `throw new EndOfGame(...)`?

Reply:
A few reason:
- ava doesn't like if we throw things that don't inherit from `Error` and I couldn't really find a way to work around it
- I need different `constructor` to be able to different between ending condition (`Win` or `Bad` word)
- it's convenient for the actions to use the promise model for this, `then()` is for the game to continue, `catch()` is for the game to end